### ISSUE TITLE
SSU regions

### ISSUE THREAD

**Issue by iferres** [ggnatalia/MMs#1]
Hi, first issue!! :raised_hand_with_fingers_splayed:  :grimacing: It's not actually an issue, tho.

I was wondering if it is possible to specify the hyper-variable regions of the 16S gene, and/or any combination of them. i.e.  V3, V4, V3-V4 (from the beginning of V3 to the end of V4), etc. If it is not possible, I would like to suggest this feature for future releases, I think it would greatly enhance the application.

Great piece of open source software btw, thanks!

**Comment by ggnatalia** [comment]
Hi!

In fact, we have included it in the latest version! You have different options. So, for example, imagine that you are interested in region V3-V4 (you should repeat and an analogous process for V3 or V4):

First, you need a tab-separated file like this one:
[V3V4.txt](https://github.com/ggnatalia/MMs/files/6894017/V3V4.txt)
Each row contains the name of the region, the start and end positions and the probability of having a mutation. 
We have chosen V3V4 positions based on the paper: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2562909/ given that it is not possible to fix these positions for each taxa. But you can find other references such as: https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-016-0992-y/tables/1

Then, as you want mutations just in those defined regions, you should distribute the mutation rate (which must sum 1) among the regions, for this reason, I have assigned 0.5 to each region. By doing this, you should find all the mutations in the defined regions.


Then, you have different alternatives:

**1)** Work with the complete sequence of the 16S rRNA gen but just mutating V3-V4:

    python3 makemocks.py  -o R_V3V4_complete -m mock1 -N 1 -nASVs 1000 -H 4 -r 10000 --cpus 20 --by_region utils/V3V4.txt

All mutations will be in the defined regions, but if you leave threshold = 0.97 (default threshold), which means that cutoff = 0.03 == 3%, all the mutations will be inside those regions, so, if the average length of the 16S is 1500 nt, if you calculate the 3% of 1500, it gives you 45 nt to be mutated in those regions. 

So, when you calculate distances (among one species and its microdiversity) using the complete sequence (all the length of the 16S ~ 1500 nt), you will find that they share > 97% of similarity, but if you just look the V3V4 regions, they share less than 97% ('cause the number of nucleotides to be mutated is calculated according the total length of the sequence).

Please, take a look to: 
[CP007533.G8DEn160.complete.align.report.txt](https://github.com/ggnatalia/MMs/files/6894810/CP007533.G8DEn160.complete.align.report.txt)
(I realign sequences agains SILVA db using mothur, you see that they share a 97% of similarity with the original sequence, but if I realign only the sequence among 6388-25319 (see point 3)), the similarity with the original sequence is lower than 97%. However, mutations are in the defined regions (sheet 1 in the excel)

To avoid this, read point 2:

**2)** Work with the complete sequence of the 16S, mutating V3V4 but using another cutoff to make microdiversity:

    python3 makemocks.py  -o R_V3V4_complete0997 -m mock1 -N 1 -nASVs 1000 -H 4 -r 10000 --cpus 20 --by_region utils/V3V4.txt --threshold 0.997 

As you can observe, I fix the threshold in 0.997 and use the same file with the defined regions.
The cutoff in this case is 0.003 == 0.3%. The 0.3% of  1500nt =  4nt to be mutated in all the 16S (but as we are defining in which regions these mutations should be, we will get 4nt mutated in the specified regions). If we compare the complete sequence of the 16S among, you will find that they share a lot of similarity, but if you compare only the specified regions, you'll find that they share > 97% of similarity.

Again, I have upload the files to look at it:
[CP013226.G8DE2250.complete0997.align.report.txt](https://github.com/ggnatalia/MMs/files/6894925/CP013226.G8DE2250.complete0997.align.report.txt)

and sheet2 of the excel


**3)**  If you don't need to work with the complete sequence of the 16S, there is another alternative, which is based on the SILVA alignment. You can select a start and end values in the alignment (for instance, V3 starts in 6388 and V4 ends in 25319 (see https://mothur.org/wiki/miseq_sop/#reducing-sequencing-and-pcr-errors). Note that in this case, we do not provide a file with the regions, you can do it, but you should define regions taking into account that positions should be referred to the trimmed sequence and not the complete 16S

I recommend you to read https://mothur.org/blog/2016/Customization-for-your-region/ to know how to define these positions in SILVA using PCR primers.

    python3 makemocks.py  -o R_V3V4 -m mock1 -N 1 -nASVs 1000 -H 4 -r 10000 --cpus 20 --start 6388 --end 25319

In this case, sequences will be trimmed between these positions, and the threshold will be applied to the resulting sequence length and not depending on the total length of the 16S. So, when you compare the final sequences, wou'll see that they share > 97% of similarity.
[CP029991.G8D30063.align.report.txt](https://github.com/ggnatalia/MMs/files/6895035/CP029991.G8D30063.align.report.txt)


**To sum up**, it depends on your purposes, but option2 may be the best: it allows you to work with the complete sequence of the 16S, but you can detect ASVs in V3/V4 :)

Excel: 
[Mutations.xls](https://github.com/ggnatalia/MMs/files/6894994/Mutations.xls)
Steps to observe this conclusions: 
[steps.txt](https://github.com/ggnatalia/MMs/files/6895097/steps.txt)


P.D. Sorry for so much text, I hope that everything is clear, but it isn't, feel free of asking!

**Comment by iferres** [comment]
Great, in the following days I'll try some of the alternatives. 
Closing this, if any related question come up I will comment here.
Thanks a lot!!

**Comment by iferres** [comment]
Hi again, question. Since it is not an issue I will keep the thread closed.

I want to compare 2 protocols: one based on Illumina V3-V4 (high accuracy), and the other based on nanopore full length 16S. So I would like to simulate **the same** reads, but for each technology so I can compare taxonomy assignments for both protocols. Based on what you explained me before I cannot understood if is it possible or what would you think it is the best approach. I want to simulate mutations on the full 16S, which I expect to use as nanopore reads (great to have the NanoSim implementation to add base qualities!), but also generate the respective V3-V4 Illumina reads.

From what I see in your explanation the case 2 is something close from what I am looking for, except it doesn't consider the nanopore reads.

**Comment by ggnatalia** [comment]
Hi! 

I do not understand exactly what you mean by simulating the same reads. One option is to use the same mock community and running a simulation for Illumina and then for Nanopore, the abundances and the reference sequences will be the same, but the reads will not (I mean, read 1 from Nanopore won't be read 1 from InSilicoSeqs but longer). Nevertheless, the read proportion  per taxa should be similar.

First create the mock community, for instance with InSilicoSeqs (Illumina):

    python3 makemocks.py  -o R_V3V4_complete0997 -m mock1 -N 1 -nASVs 1000 -H 4 -r 10000 --cpus 20 --by_region utils/V3V4.txt --threshold 0.997 

And then, repeat the read simulation but with Nanopore:

    python3 makemocks.py  -o R_V3V4_complete0997 -m mock1 -N 1 -nASVs 1000 -H 4 -r 10000 --cpus 20 --by_region utils/V3V4.txt --threshold 0.997 --Sim NanoSim --repeat_previous_mock

**Comment by iferres** [comment]
Ok, I see. Yes, I was asking if it is possible to have both illumina V3V4 and nanopore full length reads derived from exactly the same simulated 16S sequences (a 1 to 1 correspondence) in order to eliminate any source of variation when comparing the taxonomic assignments. But this is close enough, and even a more realistic approach.
Thanks!

**Comment by ggnatalia** [comment]
Ok! What do you ask it is not possible because you'd use two different sequencing simulators  (NanoSim and InSlicoSeqs), each one has its own sofware to generate reads, so there won't be the same reads. Nevertheless, in my opinion, your approach is valid given that taxa proportions would be similar.

Thanks for using M&Ms!
